### PR TITLE
First deliverable

### PR DESCRIPTION
**Pull request by Osc4ar**
## Content

This first deliverable contains the following:

- Logic to read Tasks of a ToDo List saved on a CSV
- Exposes an endpoint to get all the tasks
- Exposes an endpoint to get a task using its ID

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -0,0 +1,5 @@
+module taskmanager
+
+go 1.14
```
Comment (ivanovishado) [inline go.mod]: About the Go version: it's recommended to always use the latest one. :) As of today, that'd be 1.16

File: infrastructure/datastore/db.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+type DB interface {
+	FindByID(id uint) (*model.Task, error)
+	FindAll() []*model.Task
+}
+
+// NewDB returns a new DB instance
+func NewDB() (DB, error) {
+	csvfile, err := os.Open("tasks.csv")
+	if err != nil {
+		log.Fatalln("Could not open the CSV file", err)
```
Comment (ivanovishado) [inline infrastructure/datastore/db.go]: Statements after a `Fatal` instruction won't execute. Per [the documentation](https://golang.org/pkg/log/#Fatal):
```
The Fatal functions call os.Exit(1) after writing the log message.
```

So, I'd recommend returning the error here and handle the exit, or Fatal log, in the main function.
Please apply this comment whenever you call to a `Fatal` function.

File: infrastructure/datastore/db.go [hunk 3]
```diff
@@ -0,0 +1,110 @@
+package datastore
+
+import (
+	"encoding/csv"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"strconv"
+	"taskmanager/domain/model"
```
Comment (ivanovishado) [inline infrastructure/datastore/db.go:10]: Just a small detail: Imports tend to follow this order by convention, each section separated by a newline.
```
1. Native packages.
2. External packages.
3. Internal packages.
```

[Here](https://stackoverflow.com/a/38714480/4521547)'s an example. The `goimports` tool can help you to do this automatically.

File: infrastructure/datastore/db.go [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+		task, err := recordToTask(record)
+		if err != nil {
+			log.Fatal(err)
+			continue
+		}
+
+		tasks = append(tasks, &task)
+	}
+
+	if len(tasks) < 0 {
```
Comment (ivanovishado) [inline infrastructure/datastore/db.go]: Can this happen? Should this validation be `len(tasks) == 0` instead?

File: infrastructure/router/router.go [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+	"taskmanager/interface/controllers"
+
+	"github.com/gorilla/mux"
+)
+
+// NewRouter defines the endpoints of the Application
+func NewRouter(muxRouter *mux.Router, c controllers.AppController) *mux.Router {
+	muxRouter.HandleFunc("/tasks", func(rw http.ResponseWriter, r *http.Request) {
+		rw.Header().Add("content-type", "application/json")
+		c.GetTasks(rw)
```
Comment (ivanovishado) [inline infrastructure/router/router.go]: I guess you need to validate the error that this function returns. What do you think?

File: infrastructure/router/router.go [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+		vars := mux.Vars(r)
+
+		id, err := strconv.Atoi(vars["id"])
+		if err != nil {
+			rw.WriteHeader(http.StatusBadRequest)
+			return
+		}
+
+		rw.Header().Add("content-type", "application/json")
+		c.GetTask(uint(id), rw)
```
Comment (ivanovishado) [inline infrastructure/router/router.go]: Same comment about validating the error.

**Review by ivanovishado — APPROVED** [review]
Hi, @Osc4ar. Looking good!

Thanks for submitting your deliverable. I really liked your architecture and your code was easy to read. Kudos for that. 😄 

I would only strongly suggest that you modify the README by adding requirements, instructions to run the project, and such. Some documentation, like a Swagger file, would be really helpful as well.

I also like to recommend [this extension](https://marketplace.visualstudio.com/items?itemName=humao.rest-client) to add example calls directly in your codebase.

**Comment by Osc4ar** [comment]
Thanks for all the comments and the kudos @ivanovishado even though I submitted the deliverable delayed.

I will improve the README and also apply your comments which I think would improve a lot the code 😄

File: infrastructure/datastore/db.go (hunk 4 above)
Comment (Osc4ar) [reply infrastructure/datastore/db.go]: It cannot happen, I did not thing that condition correctly 😆 I change for `len(tasks) == 0` as you mention

File: infrastructure/datastore/db.go (hunk 3 above)
Comment (Osc4ar) [reply infrastructure/datastore/db.go:10]: Right now my VS Code setup is doing that arrangement automatically when I save, I will check that `goimports` tool to see if the arrangement is different, thanks 😄

File: infrastructure/datastore/db.go (hunk 2 above)
Comment (Osc4ar) [reply infrastructure/datastore/db.go]: You are right, I totally missed the `os.Exit(1)` implications, I am changing these to use `log.Print` and returning the error for better handling

File: infrastructure/router/router.go (hunk 5 above)
Comment (Osc4ar) [reply infrastructure/router/router.go]: Yes, I missed the case when something failed, I am adding an internal error response for these cases

File: go.mod (hunk 1 above)
Comment (Osc4ar) [reply go.mod]: I need to update my Go version then 😆 I will do that update to have it in the latest one. Thanks 👍

**Comment by ivanovishado** [comment]
No problem, @Osc4ar. Feel free to merge this one so you can send a new one for your next deliverable. 😄

**Pull request merged**